tsdb/tsm1/engine_measurement_notime_schema.go

```diff
@@ -240,3 +241,91 @@ func (e *Engine) fieldsNoTime(ctx context.Context, orgID, bucketID influxdb.ID,
 
 	return cursors.NewMeasurementFieldsSliceIterator([]cursors.MeasurementFields{{Fields: vals}}), nil
 }
+
+// MeasurementTagKeysNoTime returns an iterator which enumerates the tag keys
+// for the given bucket, measurement and tag key and filtered using the optional
+// the predicate.
+//
+// MeasurementTagKeysNoTime will always return a StringIterator if there is no error.
+//
+// If the context is canceled before MeasurementTagKeysNoTime has finished
+// processing, a non-nil error will be returned along with statistics for the
+// already scanned data.
+func (e *Engine) MeasurementTagKeysNoTime(ctx context.Context, orgID, bucketID influxdb.ID, measurement string, predicate influxql.Expr) (cursors.StringIterator, error) {
+	predicate = AddMeasurementToExpr(measurement, predicate)
+	return e.tagKeysNoTime(ctx, orgID, bucketID, predicate)
+}
+
+func (e *Engine) tagKeysNoTime(ctx context.Context, orgID, bucketID influxdb.ID, predicate influxql.Expr) (cursors.StringIterator, error) {
+	if err := ValidateTagPredicate(predicate); err != nil {
+		return nil, err
+	}
+
+	orgBucket := tsdb.EncodeName(orgID, bucketID)
+
+	vals := make([]string, 0, 128)
+
+	span := opentracing.SpanFromContext(ctx)
+	if span != nil {
+		defer func() {
+			span.LogFields(
+				log.Int("values_count", len(vals)),
+			)
+		}()
+	}
+
+	keyset := map[string]struct{}{}
+
+	if err := func() error {
+		sitr, err := e.index.MeasurementSeriesByExprIterator(orgBucket[:], predicate)
+		if err != nil {
+			return err
+		}
+		defer sitr.Close()
+
+		var tagsReuse models.Tags
+
+		for i := 0; ; i++ {
+			// to keep cache scans fast, check context every 'cancelCheckInterval' iterations
+			if i%cancelCheckInterval == 0 {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				default:
+				}
+			}
+
+			elem, err := sitr.Next()
+			if err != nil {
+				return err
+			} else if elem.SeriesID.IsZero() {
+				return nil
+			}
+
+			sf := e.index.SeriesFile()
+			if sf == nil {
+				return nil
+			}
+
+			skey := sf.SeriesKey(elem.SeriesID)
+			if len(skey) == 0 {
+				continue
+			}
+
+			_, tags := seriesfile.ParseSeriesKeyInto(skey, tagsReuse[:0])
+			for _, t := range tags {
+				keyset[string(t.Key)] = struct{}{}
+			}
```

Comment:
For the reads package, I created `KeyMerger` which you could use:

https://github.com/influxdata/influxdb/blob/0ae8bebd755cf4aa579b9c38e6ea4792dfdbe53a/storage/reads/keymerger.go#L11

Also, we can optimized`seriesfile.ParseSeriesKetInto` with a specialized implementation that just returns the keys:

```go
func parseSeriesKeys(data []byte, dst [][]byte) [][]byte {
	var name []byte
	_, data = seriesfile.ReadSeriesKeyLen(data)
	_, data = seriesfile.ReadSeriesKeyMeasurement(data)
	tagN, data := seriesfile.ReadSeriesKeyTagN(data)

	if cap(dst) < tagN {
		dst = make([][]byte, tagN)
	} else {
		dst = dst[:tagN]
	}

	for i := 0; i < tagN; i++ {
		dst[i], _, data = seriesfile.ReadSeriesKeyTag(data)
	}

	return dst
}
```

and then we can merge the result to km:

```go
keys = parseSeriesKeys(skey, keys)
km.MergeKeys(keys)
```